### ISSUE TITLE
skore-hub-project: cross validation report representation of splits

### ISSUE THREAD

**Issue by rouk1** [probabl-ai/skore#2212]
### Is your feature request related to a problem? Please describe.

Actually a cross validation report payload contains a splits filed that uses a bucket approach.
This does not work well for shuffle neither for repeating split strategy.

Random splits could be faked and kfold like split could be represented using only there "position" in the dataset.

**Comment by Premkumar-2004** [comment]
hey can i work on this issue ?

**Comment by Premkumar-2004** [comment]
Replace the bucketed distribution with the actual sorted test indices:

 Before
`"train_test_distribution": [float(np.mean(bucket)) for bucket in np.array_split(split, buckets_number)]`

After  
`"test_indices": sorted([int(i) for i in test_indices`

Why this work

->Accurate for all strategies: KFold gives contiguous ranges `[0,1,2,3,4]` ShuffleSplit gives scattered indices `[1,5,7,12,...] `both are correctly represented
->Reasonable payload size: Test sets are typically `20-50%` of data, comparable to the current 200-bucket approach
->Frontend flexibility: The UI can decide visualization (range bars for contiguous, scatter for random)
->No information loss: Exact indices vs averaged approximations

**Comment by Premkumar-2004** [comment]
hey i raised the pr for this issue please review it #2348

**Comment by thomass-dev** [comment]
Thanks for your contribution @Premkumar-2004 , but this issue have to be addressed by a maintainer, since it requires modifications on the hub side. Next time, don't hesitate to wait for feedback before starting development.

Also, your contribution looks a little too much like LLM-generated, at least your summary in this issue.
Please refer to the contributing guide https://github.com/probabl-ai/skore/blob/main/CONTRIBUTING.rst#automated-contributions-policy.

**Comment by glemaitre** [comment]
OK so looking at the problem a bit closer, I think that we can generate ahead the data to be used for the plotting. In short given the following splitter:

```python
from sklearn.model_selection import (
    GroupKFold,
    GroupShuffleSplit,
    KFold,
    ShuffleSplit,
    StratifiedGroupKFold,
    StratifiedKFold,
    StratifiedShuffleSplit,
    TimeSeriesSplit,
)
```

We can generate a fix dataset with the same property than the real dataset but only on `100` samples. Passing the class labels and groups, then we can get the index of the training and testing samples from calling the `splitter` for each split and just send those to the hub.

Also, we force this generation without any shuffling that should provide use a comprehensible graph. The repeated splitter, will reuse the exact same representation but should show an axis mentioning that we shuffle the data.

**Comment by glemaitre** [comment]
So I think that in `skore-hub-project` we should implement the following function

<details>

```python
import inspect
import numpy as np
import pandas as pd
from sklearn.model_selection import (
    KFold,
    RepeatedKFold,
    StratifiedKFold,
    RepeatedStratifiedKFold,
    ShuffleSplit,
    StratifiedShuffleSplit,
    TimeSeriesSplit,
    # ignoring the group for the moment
)

SPLITTERS = {
    "KFold": KFold,
    "RepeatedKFold": KFold,
    "StratifiedKFold": StratifiedKFold,
    "RepeatedStratifiedKFold": StratifiedKFold,
    "ShuffleSplit": ShuffleSplit,
    "StratifiedShuffleSplit": StratifiedShuffleSplit,
    "TimeSeriesSplit": TimeSeriesSplit,
}

def create_splitter_representation(splitter, target, is_classifier):
    # create metdata for the splitter
    n_repeats = getattr(splitter, "n_repeats", None)
    n_splits = splitter.get_n_splits() // (n_repeats or 1)
    splitter_metadata = {
        "type": splitter.__class__.__name__,
        "n_splits": n_splits,
        "n_repeats": n_repeats,
        # first check if shuffle is available; otherwise, we could have splitter always
        # shuffling and only exposing a random_state
        "shuffle": getattr(splitter, "shuffle", hasattr(splitter, "random_state")),
        "random_state": getattr(splitter, "random_state", None),
    }

    # create an undersampled target to create a simplify representation
    rng = np.random.default_rng(0)
    n_samples_original, n_samples_repr = len(target), 100
    if is_classifier:
        if not isinstance(target, pd.Series):
            target = pd.Series(target)
        probs = target.value_counts(normalize=True)
        print(probs)
        target_repr = rng.choice(
            probs.index.to_numpy(),  # classes
            size=n_samples_repr,
            p=probs.to_numpy(),  # probabilities
            replace=True,
        )
        target_repr.sort()
    else:  # regression
        # uniformly sample the target because it will have no impact on the
        # representation
        target_repr = rng.choice(target, size=n_samples_repr, replace=False)

    # create a simplified splitter without randomization and repetitions
    simplified_cls = SPLITTERS.get(
        splitter.__class__.__name__,
        splitter.__class__,
    )
    sig = inspect.signature(simplified_cls.__init__)
    accepted_params = set(sig.parameters.keys()) - {"self"}
    params = {}
    for name in accepted_params:
        if name in ["n_splits", "n_repeats"]:
            params[name] = splitter_metadata[name]
        elif name == "shuffle":
            params[name] = False
        elif name == "random_state":
            params[name] = None
        elif hasattr(splitter, name):
            params[name] = getattr(splitter, name)
    simplified_splitter = simplified_cls(**params)

    # create synthetic indices
    df = []
    X = rng.normal(size=(n_samples_repr, 1))
    for split_idx, (train_idx, test_idx) in enumerate(
        simplified_splitter.split(X, target_repr)
    ):
        df.append(
            pd.DataFrame(
                {
                    "split": [split_idx] * train_idx.size,
                    "dataset": ["train"] * train_idx.size,
                    "indices": train_idx,
                }
            )
        )
        df.append(
            pd.DataFrame(
                {
                    "split": [split_idx] * test_idx.size,
                    "dataset": ["test"] * test_idx.size,
                    "indices": test_idx,
                }
            )
        )
    df = pd.concat(df, axis=0, ignore_index=False)

    return {"splitter": splitter_metadata, "dataset_size": n_samples_repr, "data": df}
```

</details>

Here is a set of unit tests:

<details>

```python
# %%
# Splitters with varied parameters for testing

def _expected_n_splits(splitter):
    """n_splits for metadata: get_n_splits() // n_repeats for repeated, else get_n_splits()."""
    n_repeats = getattr(splitter, "n_repeats", None)
    return splitter.get_n_splits() // (n_repeats or 1)


def _assert_representation_structure(result, splitter, expected_n_splits):
    """Basic asserts on the dictionary returned by create_splitter_representation."""
    meta = result["splitter"]
    assert meta["type"] == splitter.__class__.__name__
    assert meta["n_splits"] == expected_n_splits
    assert meta["n_repeats"] == getattr(splitter, "n_repeats", None)
    assert "shuffle" in meta
    assert "random_state" in meta
    assert result["dataset_size"] == 100
    df = result["data"]
    assert isinstance(df, pd.DataFrame)
    assert set(df.columns) == {"split", "dataset", "indices"}
    assert set(df["dataset"].unique()) == {"train", "test"}
    n_splits_actual = df["split"].nunique()
    assert n_splits_actual == expected_n_splits, (n_splits_actual, expected_n_splits)


# %%
def test_repeated_splitters():
    """Test RepeatedKFold and RepeatedStratifiedKFold metadata and structure."""
    configs = [
        (
            "RepeatedKFold",
            RepeatedKFold(n_splits=3, n_repeats=2, random_state=0),
            False,
        ),
        (
            "RepeatedKFold_other",
            RepeatedKFold(n_splits=5, n_repeats=3, random_state=42),
            False,
        ),
        (
            "RepeatedStratifiedKFold",
            RepeatedStratifiedKFold(n_splits=4, n_repeats=2, random_state=1),
            True,
        ),
    ]
    for name, splitter, is_classifier in configs:
        target = labels if is_classifier else rng.normal(size=n_samples)
        result = create_splitter_representation(splitter, target, is_classifier)
        expected_n_splits = _expected_n_splits(splitter)
        _assert_representation_structure(result, splitter, expected_n_splits)
        assert result["splitter"]["n_repeats"] == splitter.n_repeats


def test_stratified_splitters():
    """Test StratifiedKFold, RepeatedStratifiedKFold, StratifiedShuffleSplit."""
    configs = [
        (
            "StratifiedKFold",
            StratifiedKFold(n_splits=3, shuffle=True, random_state=0),
            True,
        ),
        (
            "StratifiedKFold_no_shuffle",
            StratifiedKFold(n_splits=4, shuffle=False),
            True,
        ),
        (
            "RepeatedStratifiedKFold",
            RepeatedStratifiedKFold(n_splits=3, n_repeats=2, random_state=1),
            True,
        ),
        (
            "StratifiedShuffleSplit",
            StratifiedShuffleSplit(n_splits=5, random_state=2),
            True,
        ),
    ]
    for name, splitter, is_classifier in configs:
        result = create_splitter_representation(splitter, labels, is_classifier=True)
        expected_n_splits = _expected_n_splits(splitter)
        _assert_representation_structure(result, splitter, expected_n_splits)
        assert result["splitter"]["type"] in (
            "StratifiedKFold",
            "RepeatedStratifiedKFold",
            "StratifiedShuffleSplit",
        )

def test_other_splitters():
    """Test KFold, ShuffleSplit, TimeSeriesSplit (non-stratified, non-repeated)."""
    configs = [
        ("KFold", KFold(n_splits=3), False),
        ("KFold_shuffle", KFold(n_splits=5, shuffle=True, random_state=10), False),
        ("ShuffleSplit", ShuffleSplit(n_splits=4, random_state=20), False),
        ("TimeSeriesSplit", TimeSeriesSplit(n_splits=3), False),
    ]
    for name, splitter, is_classifier in configs:
        target = labels if is_classifier else rng.normal(size=n_samples)
        result = create_splitter_representation(splitter, target, is_classifier)
        _assert_representation_structure(result, splitter, _expected_n_splits(splitter))
        assert result["splitter"]["n_repeats"] is None
```


</details>

**Comment by glemaitre** [comment]
@thomass-dev I would need to incorporate the changes in here:

https://github.com/probabl-ai/skore/blob/0abd763a482e4795e3d4e0f3ef8d51ab8b44cb08/skore-hub-project/src/skore_hub_project/report/cross_validation_report.py#L205-L228

I assume that we will need a migration of the database model.
